docs/azure/sdk/unit-testing-mocking.md

```diff
@@ -0,0 +1,420 @@
+---
+title: Unit testing and mocking with the Azure SDK .NET
+description: Learn techniques and tools for unit testing and mocking the Azure SDK for .NET
+ms.custom: devx-track-dotnet, engagement-fy23
+ms.date: 07/05/2023
+---
+
+# Unit testing and mocking with the Azure SDK for .NET
+
+Unit testing is an important part of a sustainable development process that can improve code quality and prevent regressions or bugs in your apps. However, unit testing presents challenges when the code you're testing performs network calls to Azure resources. Tests that run against live services can experience issues, such as, latency that slows down test execution, dependencies on code outside of the isolated test, and issues with managing service state and costs every time the test is run. Instead of testing against live Azure services, replace the service clients with mocked or in-memory implementations. This avoids the above issues and lets developers focus on testing their application logic, independent from the network and service.
+
+In this article, you'll learn how to write unit tests for the Azure SDK for .NET that isolate your dependencies to make your tests more reliable. You'll also learn how to replace key components with in-memory test implementations to create fast and reliable unit tests, and see how to design your own classes to better support unit testing. This article includes examples that use [Moq](https://www.nuget.org/packages/moq/), which is a popular mocking and testing library for .NET.
+
+## Understand service clients
+
+A service client class is the main entry point for developers in an Azure SDK library and implements most of the logic to communicate with the Azure service. When unit testing service client classes, it's important to be able to create an instance of the client that behaves as expected without making any network calls.
+
+Each of the Azure SDK clients follows mocking guidelines that allow their behavior to be overridden:
+
+* Each client offers at least one protected constructor to allow inheritance for testing.
+* All public client members are virtual to allow overriding.
+
+> [!NOTE]
+> The code examples in this article use types from the [`Azure.Security.KeyVault.Secrets`](https://www.nuget.org/packages/Azure.Security.KeyVault.Secrets/) library for the Azure Key Vault service. The concepts demonstrated in this article also apply to service clients from many other Azure services, such as Azure Storage or Azure Service Bus.
+
+To create a test service client, you can either use a mocking library such as [Moq](https://www.nuget.org/packages/moq/) or standard C# features such as inheritance. Mocking frameworks allow you to simplify the code that you must write to override member behavior (as well as other useful features that are beyond the scope of this article).
```

Comment:
💡 Mocking libraries like Moq are especially harmful to the health of a test suite when used at scale. Since they do not follow the inheritance rules defined by the language, features such as Go To Implementation will not correctly identify them as implementations of a member when developers are working to update their code. This can lead to large numbers of false positive test failures in non-breaking development scenarios.

Reply:
➡️ Absorbed by https://github.com/dotnet/docs/pull/36110#discussion_r1255846092